blocks/main_loop.py

```diff
@@ -42,6 +45,87 @@
 a model. Check documentation of your extensions."""
 
 
+class Timer(object):
```

Comment:
It seems to me that this is class is generic and not specific to the main loop and should go to a separate file. Also it needs a bit of docs to explain how it's different from other classes named `Timer`, e.g. `timeit.Timer` from the standard library. If I get right, it's special and very nice feature is that it's hierarchical, maybe it should be even called `HierachicalTimer`?